### PR TITLE
Handle git dependencies, roll packages to get transitive deps of flutter_gallery

### LINKED ISSUE #61044: The new flutter gallery has unpinned depdendencies, and macrobenchmarks depends on flutter gallery
See https://github.com/flutter/flutter/blob/fb57edcc80dbfc75e2f42e4fb52ff730c13c29c8/dev/benchmarks/macrobenchmarks/pubspec.yaml#L21-L24

And https://github.com/flutter/gallery/blob/8ec7fdb5d599f8fd085b1d18de6f24a765fd07af/pubspec.yaml#L9

Causes errors like: 
```
▌20:52:15▐ RUNNING: cd .; bin/flutter analyze --dartdocs --flutter-repo
Package "url_launcher" has conflicts:
  1 source wants "/root/.pub-cache/hosted/pub.dartlang.org/url_launcher-5.5.0/lib":
    /tmp/flutter sdk/dev/benchmarks/macrobenchmarks/.packages
  1 source wants "/root/.pub-cache/hosted/pub.dartlang.org/url_launcher-5.4.11/lib":
    /tmp/flutter sdk/dev/integration_tests/flutter_gallery/.packages
Make sure you have run "pub upgrade" in all the directories mentioned above.
```

When a plugin like url_launcher publishes a new version that the Flutter repo hasn't be rolled to yet.

I don't think we can fix this without removing the Gallery as a dependency on macrobenchmarks...

@yjbanov @jonahwilliams @zanderso

### PR DESCRIPTION
**Pull request by dnfield**
Fixes https://github.com/flutter/flutter/issues/61044

Handles git dependencies (in a slightly ugly way - can refactor if this is really a problem) in the `update-packages --force-upgrade` logic. 

This prevents `gallery` from giving `macrobenchmarks` arbitrary deps higher than what we expect for the repo.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/update_packages.dart [hunk 1]
```diff
@@ -644,23 +648,23 @@ class PubspecYaml {
...
-          result.add(PubspecLine(line));
+          // Ok we're dealing with some "git:" dependency. Consume lines until
+          // we are out of the git dependency, and stuff them into the lock
+          // line.
+          lastDependency._lockLine = line;
+          lastDependency._lockIsOverride = section == Section.dependencyOverrides;
+          do {
+            index += 1;
+            line = lines[index];
+            lastDependency._lockLine += '\n$line';
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/commands/update_packages.dart:666]: I could make a method on `PubspecDependency` to handle this, but  .. this whole thing is pretty convoluted as it is, and all the classes are in this file anyway.

**Review by jonahwilliams — APPROVED** [review]
If it works, it works

LGTM

File: packages/flutter_tools/lib/src/commands/update_packages.dart [hunk 2]
```diff
@@ -644,23 +648,23 @@ class PubspecYaml {
...
-          result.add(PubspecLine(lastDependency.line));
-          result.add(PubspecLine(line));
+          // Ok we're dealing with some "git:" dependency. Consume lines until
+          // we are out of the git dependency, and stuff them into the lock
+          // line.
+          lastDependency._lockLine = line;
+          lastDependency._lockIsOverride = section == Section.dependencyOverrides;
+          do {
+            index += 1;
+            line = lines[index];
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/update_packages.dart:665]: It seems like there should be a bounds check on `lines.length` somewhere in this loop.

**Review by zanderso — COMMENTED** [review]
dbc

File: packages/flutter_tools/lib/src/commands/update_packages.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter_tools/lib/src/commands/update_packages.dart:665]: Yeah, an invalid pubspec could cause this to crash. I'll add a check and throw with a more appropriate error.

**Pull request merged**